File: 1311-LargestMagicSquare/1311-LargestMagicSquare.cpp
// Last updated: 4/3/2026, 1:57:38 PM
class Solution {
public:
    int largestMagicSquare(vector<vector<int>>& grid) {
        int m = grid.size(), n = grid[0].size();
        vector<vector<int>> rows(m + 1, vector<int>(n + 1, 0));
        vector<vector<int>> cols(m + 1, vector<int>(n + 1, 0));

        vector<vector<int>> diag1(m + 1, vector<int>(n + 1, 0));
        vector<vector<int>> diag2(m + 1, vector<int>(n + 2, 0));

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                rows[i + 1][j + 1] = rows[i + 1][j] + grid[i][j];
                cols[i + 1][j + 1] = cols[i][j + 1] + grid[i][j];
                diag1[i + 1][j + 1] = diag1[i][j] + grid[i][j];
                diag2[i + 1][j + 1] = diag2[i][j + 2] + grid[i][j];
            }
        }
        for (int k = min(m, n); k > 0; k--) {
            for (int i = 0; i + k <= m; i++) {
                for (int j = 0; j + k <= n; j++) {
                    int sum = rows[i + 1][j + k] - rows[i + 1][j];
                    bool flag = true;

                    for (int r = 0; r < k && flag; r++)
                        flag &= (rows[i + r + 1][j + k] - rows[i + r + 1][j] ==
                                 sum);
                    for (int c = 0; c < k && flag; c++)
                        flag &= (cols[i + k][j + c + 1] - cols[i][j + c + 1] ==
                                 sum);

                    flag &= (diag1[i + k][j + k] - diag1[i][j] == sum);
                    flag &= (diag2[i + k][j + 1] - diag2[i][j + k + 1] == sum);
                    if (flag)
                        return k;
                }
            }
        }
        return 1;
    }
};

File: 1311-LargestMagicSquare/1311-LargestMagicSquare.py
# Last updated: 4/3/2026, 1:57:33 PM
class Solution:
    def largestMagicSquare(self, grid: List[List[int]]) -> int:
        m, n = len(grid), len(grid[0])
        rows = [[0 for _ in range(n+1)] for _ in  range(m+1)]
        cols = [[0 for _ in range(n+1)] for _ in  range(m+1)]  
        diag1 = [[0 for _ in range(n+1)] for _ in  range(m+1)]
        diag2 = [[0 for _ in range(n+2)] for _ in  range(m+1)]

        for i in range(m):
            for j in range(n):
                rows[i+1][j+1] = rows[i+1][j] + grid[i][j]
                cols[i+1][j+1] = cols[i][j+1] + grid[i][j]

                diag1[i+1][j+1] = diag1[i][j] + grid[i][j]
                diag2[i+1][j+1] = diag2[i][j+2] + grid[i][j]
        
        for k in range(min(m,n), 0, -1):
            for i in range(m-k+1):
                for j in range(n-k+1):
                    sum = rows[i+1][j+k] - rows[i+1][j]
                    flag = True

                    for r in range(k):
                        if not flag:
                            break
                        flag = flag and (rows[i+r+1][j+k]-rows[i+r+1][j] == sum)
                    
                    for c in range(k):
                        if not flag:
                            break
                        flag = flag and(cols[i+k][j+c+1]-cols[i][j+c+1] == sum)
                    
                    flag = flag and (diag1[i+k][j+k] - diag1[i][j] == sum)
                    flag = flag and (diag2[i+k][j+1] - diag2[i][j+k+1] == sum)

                    if flag:
                        return k
        return 1